xla_native_functions.yaml

```diff
@@ -317,6 +317,8 @@ supported:
   - lerp.Scalar
   - lerp.Tensor
   - logical_and.out
+  - adaptive_max_pool2d.out
```

Comment:
Actually, I think you're better off implementing the functional (out-of-place) versions of both of these operators instead of the `out=` overloads. Since with the new codegen, we'll generate the inplace and out= variants for you based on your functional implementations.

This will work too, but it'll be slightly less efficient. If you're curious: `adaptive_max_pool2d` and `adaptive_max_pool2d` are both implemented as structured kernels in core (see [here](https://github.com/pytorch/pytorch/blob/979180cd0118f9379680df9e39fc6acb80d7b80a/aten/src/ATen/native/native_functions.yaml#L8769)), which means that we generate composite implementations of the out-of-place operator. It's implemented by calling into the dispatcher to create an appropriately-sized empty tensor, and calling into the dispatcher again to call the `out=` operator (which will call your `out=` xla lowering).

That'll all work fine, but results in a little bit of unnecessary overhead. It should be easy to lower the out-of-place versions instead - just replace `adaptive_max_pool2d.out` with `adaptive_max_pool2d` (same with the backward), and remove the `out` argument from the signature.


I know that there are a bunch of other ops that are also lowered as `out=` kernels. We can probably reduce a bit of framework overhead by changing them to be out-of-place lowerings instead, but I don't think it's pressing. I can add a PR for that later.

Reply:
Thanks @bdhirsh , I didn't know this and we should put this in `OP_LOWERING_GUIDE` too 😄

Reply:
@bdhirsh So in the future I should also try to implement the op with `structured_delegate` entry right?

Reply:
> I didn't know this and we should put this in OP_LOWERING_GUIDE too

Yeah - the example you have in the lowering guide actually does that, but it could probably be more explicit. I'll add a PR.


> So in the future I should also try to implement the op with structured_delegate entry right?

Technically, you might also see an inplace op with the `structured_delegate` keyword, but that will also have a trailing underscore in the name. I also don't think that you need the operator to be structured in order to do this, since the only requirement in the external codegen is that it's able to find a matching out-of-place and `out=` operator that it can group together (for example, `add.Tensor` and `add.out`). The way that I would check is:
- For a given operator you want to lower, look at all the existing overloads in `native_functions.yaml`, and find the version that doesn't take an `out` argument. It's usually the version that doesn't have an overload name (like in this case, with `adaptive_max_pool2d`), although that's not always true (for torch.add, the op name is `add.Tensor`, not just `add`). As long as there's also a corresponding `out=` operator with the same operator name and signature, but with an extra out argument, then the codegen will automatically generate the corresponding `out=` lowering for you